vault/plugin_catalog.go

```diff
@@ -34,20 +39,135 @@ type PluginCatalog struct {
 	lock sync.RWMutex
 }
 
-func (c *Core) setupPluginCatalog() error {
+func (c *Core) setupPluginCatalog(ctx context.Context) error {
 	c.pluginCatalog = &PluginCatalog{
 		builtinRegistry: c.builtinRegistry,
 		catalogView:     NewBarrierView(c.barrier, pluginCatalogPath),
 		directory:       c.pluginDirectory,
 	}
 
+	err := c.pluginCatalog.UpgradePlugins(ctx, c.logger)
+	if err != nil {
+		c.logger.Error("error while upgrading plugin storage", "error", err)
+	}
+
 	if c.logger.IsInfo() {
 		c.logger.Info("successfully setup plugin catalog", "plugin-directory", c.pluginDirectory)
 	}
 
 	return nil
 }
 
+func (c *PluginCatalog) UpgradePlugins(ctx context.Context, logger log.Logger) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// If the directory isn't set we can skip the upgrade attempt
+	if c.directory == "" {
+		return nil
+	}
+
+	// List plugins from old location
+	pluginsRaw, err := c.catalogView.List(ctx, "")
+	if err != nil {
+		return err
+	}
+	plugins := make([]string, 0, len(pluginsRaw))
```

Comment:
Since we're `append`ing, we might as well initialize it with 0 length so that we don't range over empty values down later. Otherwise the `plugins` slice may contain up to 3 extra empty strings.

Reply:
i'm actually setting the capacity, so the length will always be correct, we might just sometimes have a bit more capacity than we need

Reply:
Ahh right, the third arg is for capacity. Can we change it to `cap(pluginsRaw)`?

Reply:
I don't think that's necessary, the capacity if pluginsRaw isn't important here. We only care about the length